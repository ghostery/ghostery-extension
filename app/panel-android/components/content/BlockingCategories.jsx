/**
 * Blocking Categories Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';
import BlockingCategory from './BlockingCategory';

class BlockingCategories extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			openCategoryIndex: -1,
			blockingType: props.type,
		};
	}

	static getDerivedStateFromProps(props, state) {
		const { type } = props;
		const { blockingType } = state;

		if (type !== blockingType) {
			return {
				openCategoryIndex: -1,
				blockingType: type,
			};
		}
		return null;
	}

	getOpenStatus = (index) => {
		const { openCategoryIndex } = this.state;
		return index === openCategoryIndex;
	}

	toggleCategoryOpen = (index) => {
		const { openCategoryIndex } = this.state;
		if (openCategoryIndex === index) {
			this.setState({ openCategoryIndex: -1 });
		} else {
			this.setState({ openCategoryIndex: index });
		}
	}

	render() {
		const {
			categories,
			type,
			siteProps,
			settings,
			callGlobalAction,
		} = this.props;

		return (
			<div className="BlockingCategories">
				{
					categories.map((category, index) => (
						<BlockingCategory
							key={category.id}
							index={index}
							category={category}
							settings={settings}
							toggleCategoryOpen={this.toggleCategoryOpen}
							open={this.getOpenStatus(index)}
							type={type}
							siteProps={siteProps}
							callGlobalAction={callGlobalAction}
						/>
					))
				}
			</div>
		);
	}
}

BlockingCategories.propTypes = {
	categories: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	type: PropTypes.oneOf([
		'site',
		'global',
	]).isRequired,
	siteProps: PropTypes.shape({}).isRequired,
	callGlobalAction: PropTypes.func.isRequired,
	settings: PropTypes.shape({}).isRequired,
};

export default BlockingCategories;
