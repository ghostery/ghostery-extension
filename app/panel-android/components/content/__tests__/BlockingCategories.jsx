/**
 * BlockingCategories Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import BlockingCategories from '../BlockingCategories';

describe('app/panel-android/components/content/BlockingCategories.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('BlockingCategories component as site', () => {
			const categories = [
				{
					id: 'cat-1',
					name: 'Category-1',
					num_total: 1,
					num_blocked: 1,
					trackers: [
						{
							id: 1,
							name: 'Tracker 1',
							ss_allowed: false,
							ss_blocked: false,
							blocked: true,
						},
					],
					img_name: 'category-1-image-url',
				},
				{
					id: 'cat-2',
					name: 'Category-2',
					num_total: 5,
					num_blocked: 3,
					trackers: [
						{
							id: 2,
							name: 'Tracker 2',
							ss_allowed: false,
							ss_blocked: false,
							blocked: true,
						},
						{
							id: 3,
							name: 'Tracker 3',
							ss_allowed: false,
							ss_blocked: false,
							blocked: true,
						},
						{
							id: 4,
							name: 'Tracker 4',
							ss_allowed: false,
							ss_blocked: false,
							blocked: true,
						},
						{
							id: 5,
							name: 'Tracker 5',
							ss_allowed: false,
							ss_blocked: false,
							blocked: false,
						},
						{
							id: 6,
							name: 'Tracker 6',
							ss_allowed: false,
							ss_blocked: false,
							blocked: false,
						},
					],
					img_name: 'category-2-image-url',
				},
			];
			const siteProps = {
				isTrusted: false,
				isRestricted: false,
				isPaused: false,
			};

			const component = renderer.create(
				<BlockingCategories
					type="site"
					categories={categories}
					settings={{}}
					siteProps={siteProps}
					callGlobalAction={() => {}}
				/>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('BlockingCategories component as global', () => {
			const categories = [
				{
					id: 'cat-1',
					name: 'Category-1',
					num_total: 1,
					num_blocked: 1,
					trackers: [
						{
							id: '1',
							name: 'Tracker 1',
							blocked: true,
						},
					],
					img_name: 'category-1-image-url',
				},
				{
					id: 'cat-2',
					name: 'Category-2',
					num_total: 5,
					num_blocked: 3,
					trackers: [
						{
							id: '2',
							name: 'Tracker 2',
							blocked: true,
						},
						{
							id: '3',
							name: 'Tracker 3',
							blocked: true,
						},
						{
							id: '4',
							name: 'Tracker 4',
							blocked: true,
						},
						{
							id: '5',
							name: 'Tracker 5',
							blocked: false,
						},
						{
							id: '6',
							name: 'Tracker 6',
							blocked: false,
						},
					],
					img_name: 'category-2-image-url',
				},
			];
			const siteProps = {
				isTrusted: false,
				isRestricted: false,
				isPaused: false,
			};

			const component = renderer.create(
				<BlockingCategories
					type="global"
					categories={categories}
					settings={{}}
					siteProps={siteProps}
					callGlobalAction={() => {}}
				/>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Functionality tests shallow mounted with Enzyme', () => {
		test('BlockingCategories component toggle category clicks work', () => {
			const categories = [
				{
					id: 'cat-1',
					name: 'Category-1',
					num_total: 1,
					num_blocked: 1,
					trackers: [
						{
							id: '1',
							name: 'Tracker 1',
							blocked: true,
						},
					],
					img_name: 'category-1-image-url',
				},
				{
					id: 'cat-2',
					name: 'Category-2',
					num_total: 5,
					num_blocked: 3,
					trackers: [
						{
							id: '2',
							name: 'Tracker 2',
							blocked: true,
						},
						{
							id: '3',
							name: 'Tracker 3',
							blocked: true,
						},
						{
							id: '4',
							name: 'Tracker 4',
							blocked: true,
						},
						{
							id: '5',
							name: 'Tracker 5',
							blocked: false,
						},
						{
							id: '6',
							name: 'Tracker 6',
							blocked: false,
						},
					],
					img_name: 'category-2-image-url',
				},
			];
			const siteProps = {
				isTrusted: false,
				isRestricted: false,
				isPaused: false,
			};

			const component = shallow(
				<BlockingCategories
					type="site"
					categories={categories}
					settings={{}}
					siteProps={siteProps}
					callGlobalAction={() => {}}
				/>
			);
			const instance = component.instance();

			expect(component.state('openCategoryIndex')).toBe(-1);
			expect(instance.getOpenStatus(0)).toBe(false);

			instance.toggleCategoryOpen(0);
			expect(component.state('openCategoryIndex')).toBe(0);
			expect(instance.getOpenStatus(0)).toBe(true);

			component.setProps({ type: 'global' });
			expect(component.state('openCategoryIndex')).toBe(-1);
			expect(instance.getOpenStatus(0)).toBe(false);
		});
	});
});
