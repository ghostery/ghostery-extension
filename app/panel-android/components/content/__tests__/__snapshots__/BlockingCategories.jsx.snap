// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel-android/components/content/BlockingCategories.jsx Snapshot tests with react-test-renderer BlockingCategories component as global 1`] = `
<div
  className="BlockingCategories"
>
  <div
    className="BlockingCategory"
  >
    <div
      className="BlockingCategory__details flex-container"
      onClick={[Function]}
    >
      <img
        className="BlockingCategory__image"
        src="/app/images/panel-android/categories/category-1-image-url.svg"
      />
      <div
        className="flex-child-grow"
      >
        <h2
          className="BlockingCategory__name"
        >
          Category-1
        </h2>
        <div>
          <span
            className="BlockingCategory__numTrackers"
          >
            1 blocking_category_tracker
          </span>
          <span
            className="BlockingCategory__numBlocked"
          >
            1 blocking_category_blocked
          </span>
        </div>
      </div>
      <div
        className="BlockingCategory__buttons flex-container flex-dir-column align-justify align-middle"
      >
        <div
          className="BlockingCategory--noPointer"
        >
          <div
            className="BlockingSelectButton BlockingSelectButton__blocked"
            onClick={[Function]}
          />
        </div>
        <div
          className="BlockingCategory__toggle"
        />
      </div>
    </div>
    <div
      className="BlockingCategory__list"
      style={
        Object {
          "height": 0,
        }
      }
    />
  </div>
  <div
    className="BlockingCategory"
  >
    <div
      className="BlockingCategory__details flex-container"
      onClick={[Function]}
    >
      <img
        className="BlockingCategory__image"
        src="/app/images/panel-android/categories/category-2-image-url.svg"
      />
      <div
        className="flex-child-grow"
      >
        <h2
          className="BlockingCategory__name"
        >
          Category-2
        </h2>
        <div>
          <span
            className="BlockingCategory__numTrackers"
          >
            5 blocking_category_trackers
          </span>
          <span
            className="BlockingCategory__numBlocked"
          >
            3 blocking_category_blocked
          </span>
        </div>
      </div>
      <div
        className="BlockingCategory__buttons flex-container flex-dir-column align-justify align-middle"
      >
        <div
          className="BlockingCategory--noPointer"
        >
          <div
            className="BlockingSelectButton BlockingSelectButton__mixed"
            onClick={[Function]}
          />
        </div>
        <div
          className="BlockingCategory__toggle"
        />
      </div>
    </div>
    <div
      className="BlockingCategory__list"
      style={
        Object {
          "height": 0,
        }
      }
    />
  </div>
</div>
`;

exports[`app/panel-android/components/content/BlockingCategories.jsx Snapshot tests with react-test-renderer BlockingCategories component as site 1`] = `
<div
  className="BlockingCategories"
>
  <div
    className="BlockingCategory"
  >
    <div
      className="BlockingCategory__details flex-container"
      onClick={[Function]}
    >
      <img
        className="BlockingCategory__image"
        src="/app/images/panel-android/categories/category-1-image-url.svg"
      />
      <div
        className="flex-child-grow"
      >
        <h2
          className="BlockingCategory__name"
        >
          Category-1
        </h2>
        <div>
          <span
            className="BlockingCategory__numTrackers"
          >
            1 blocking_category_tracker
          </span>
          <span
            className="BlockingCategory__numBlocked"
          >
            1 blocking_category_blocked
          </span>
        </div>
      </div>
      <div
        className="BlockingCategory__buttons flex-container flex-dir-column align-justify align-middle"
      >
        <div
          className="BlockingCategory--noPointer"
        >
          <div
            className="BlockingSelectButton BlockingSelectButton__blocked"
            onClick={[Function]}
          />
        </div>
        <div
          className="BlockingCategory__toggle"
        />
      </div>
    </div>
    <div
      className="BlockingCategory__list"
      style={
        Object {
          "height": 0,
        }
      }
    />
  </div>
  <div
    className="BlockingCategory"
  >
    <div
      className="BlockingCategory__details flex-container"
      onClick={[Function]}
    >
      <img
        className="BlockingCategory__image"
        src="/app/images/panel-android/categories/category-2-image-url.svg"
      />
      <div
        className="flex-child-grow"
      >
        <h2
          className="BlockingCategory__name"
        >
          Category-2
        </h2>
        <div>
          <span
            className="BlockingCategory__numTrackers"
          >
            5 blocking_category_trackers
          </span>
          <span
            className="BlockingCategory__numBlocked"
          >
            3 blocking_category_blocked
          </span>
        </div>
      </div>
      <div
        className="BlockingCategory__buttons flex-container flex-dir-column align-justify align-middle"
      >
        <div
          className="BlockingCategory--noPointer"
        >
          <div
            className="BlockingSelectButton BlockingSelectButton__mixed"
            onClick={[Function]}
          />
        </div>
        <div
          className="BlockingCategory__toggle"
        />
      </div>
    </div>
    <div
      className="BlockingCategory__list"
      style={
        Object {
          "height": 0,
        }
      }
    />
  </div>
</div>
`;
